Gutenboarding: Persisted data should expire and be conditional

jeraldjuice opened the pull request:
### Summary

This PR adds smoother handling for persisted data in the `onboard` data stores. Specifically:

- [x]  Data is only persisted for 7 days.
- [ ]  Data is not persisted during a support session.
- [x]  Persisted data is cleared with a `fresh` URL param.

<hr>

### Specifics

This is fun because there's not exactly a smooth nor clear solution.

The persistence plugin alone is not quite a quick solution here. Predominately, because creating a new storage implementation involves rewriting the default storage option that already exists for `localStorage` - albeit, encapsulated and untouchable - and serializing/re-serializing data twice (this is handled at a higher layer within the plugin and not exposed, so data handling requires a second transform), to ultimately end up doing almost the same thing. Seems like the persistence plugin was more intended for different storage implementations, like `sessionStorage`, rather than overt data handling.

Ultimately, I went with using a secondary key in `localStorage` to act as the timestamp, given that two `setItem` calls are more performant than double serializing and that it's less overtly doing something twice. Still YAPSS (Yet Another Persistent Storage Solution), but it works with the least amount of smelliness.

<hr>

### Testing

* Spin this up locally and go to `/gutenboarding`.
    * Ensure that data persists normally:
        * Pick a vertical and enter a site title, then refresh the page. Ensure that the site title and vertical are still filled out correctly.
    * Test expired persisted data flow:
         * Go to `/`, the root of Calypso. (You want to leave the `/gutenboarding` page, as interacting with the page to refresh it will undo what you are about to do.)
         * Follow the instructions in the section below to edit local storage and set the `WP_ONBOARD_TS` to `0`.
         * Go back to `/gutenboarding` and ensure that the data entered from the first few steps is no longer persisted and that it displays as a blank slate.
    * Test URL params:
         * Go to `/gutenboarding` and set a site title or other data.
         * Go to `/gutenboarding?fresh` and ensure that the persisted data is not visible.

**Interacting with localStorage through DevTools:**
* Open up DevTools and go to the "Application" tab.
* In the sidebar, under "Storage > Local Storage", select the URL of the Calypso instance you're interacting with (either your local install `https://calypso.localhost...` or a different URL).
* Find the `WP_ONBOARD_TS` entry and set it to `0`.
<img width="987" alt="Screen Shot 2020-02-07 at 2 26 29 AM" src="https://user-images.githubusercontent.com/8892849/74009829-0a77f100-4952-11ea-85d3-2faba9b4ce2a.png">


<hr>

Fixes #37849

jeraldjuice commented on client/landing/gutenboarding/stores/onboard/index.ts:
```diff
@@ -3,6 +3,7 @@
  */
 import { controls } from '@wordpress/data-controls';
 import { plugins, registerStore, use } from '@wordpress/data';
+import { isSupportSession } from 'lib/user/support-user-interop';
```
Not sure if this is the type of thing that we want to avoid importing? In our quest for an eventual solo app.

simison replied:
Good question! Since it's a `lib/` it might be quite safe to import now like this as long as we make sure it'll eventually get moved to `packages/`. I'm saying might be safe because libs tend to be quite isolated or at least small.

It does import these things:

https://github.com/Automattic/wp-calypso/blob/a1c0a829f3fc9b52820e03b06604d04d2c94ba85/client/lib/user/support-user-interop.js#L1-L15

jeraldjuice commented:
So, wrestling a bit with `isSupportSession`. Not quite a drop-and-go solution like I thought. I've removed it for now for testing purposes - I'll look more into it tomorrow.

sirreal commented:
> So, wrestling a bit with `isSupportSession`. Not quite a drop-and-go solution like I thought. I've removed it for now for testing purposes - I'll look more into it tomorrow.

I don't believe Gutenboarding makes any use of the Calypso User —`client/lib/user` or any Calypso redux (user) state— I don't think SSP is something we need to consider at all at this time.

simison commented:
> I don't believe Gutenboarding makes any use of the Calypso User —client/lib/user or any Calypso redux (user) state— I don't think SSP is something we need to consider at all at this time.

👍 Let's log this in a separate issue and follow up later.

Addison-Stavlo commented on client/landing/gutenboarding/stores/onboard/index.ts:
```diff
@@ -3,6 +3,7 @@
  */
 import { controls } from '@wordpress/data-controls';
 import { plugins, registerStore, use } from '@wordpress/data';
+//import { isSupportSession } from 'lib/user/support-user-interop';
```
> 👍 Let's log this in a separate issue and follow up later.

Agree with @sirreal  and @simison on this.

Addison-Stavlo commented on client/landing/gutenboarding/stores/onboard/persist.ts:
```diff
@@ -0,0 +1,79 @@
+const storageHandler = localStorageSupport() ? window.localStorage : objStorage;
+
+// Persisted data expires after seven days
+const isFresh = ( timestampStr: string ): boolean => {
```
Small nitpick on naming.  `isFresh` seems a bit more ambiguous than `isExpired` or `isNotExpired`.  `isFresh` makes me think we are loading on a fresh state, similar to when the `?fresh` param is used to clear the data.  But here `isFresh` is `true` when the user comes back after 4 days to pick up where they left off.

Addison-Stavlo commented on client/landing/gutenboarding/stores/onboard/persist.ts:
```diff
@@ -0,0 +1,79 @@
+	getItem( key ) {
+		const timestamp = storageHandler.getItem( STORAGE_TS_KEY );
+
+		if ( timestamp && isFresh( timestamp ) && ! hasFreshParam() ) {
```
Just noting `isFresh` seems a bit more confusing when set next to `hasFreshParam()`.
Something like `timestamp && isNotExpired( timestamp ) && ! hasFreshParam()` may be more easily readable to a dev looking through the code for the first time.

jeraldjuice commented:
Cleaned up and removed `isSupportSession`, because it looks like we may not need this here, based on @sirreal 's comment. I'll open up another issue to track that.

jeraldjuice commented:
rebased for tests to clear

sirreal commented on client/landing/gutenboarding/stores/onboard/persist.ts:
```diff
@@ -0,0 +1,79 @@
+
+// Check for "fresh" query param
+const hasFreshParam = (): boolean => {
+	return window.location.search.indexOf( 'fresh' ) !== -1;
```
This is a bit loose, matching both of these:
- `https://example.com/?not-fresh`
- `https://example.com/?not-interested=fresh`

How about the following to better cover edge cases?

```suggestion
	return new URLSearchParams( window.location.search ).has( 'fresh' )
```
